torch/csrc/lazy/core/lazy_mode.cpp

```diff
@@ -0,0 +1,75 @@
+#include <torch/csrc/lazy/core/lazy_mode.h>
+
+#include <c10/core/DispatchKey.h>
+#include <c10/core/impl/LocalDispatchKeySet.h>
+#include <torch/csrc/lazy/backend/backend_device.h>
+#include <torch/csrc/lazy/core/lazy_graph_executor.h>
+#include <torch/library.h>
+
+// TODO(whc) we can't actually depend on ts backend code from here, but we could refactor if reusing the callback
+// turns out to be the best way to implement this
+#include <torch/csrc/lazy/ts_backend/ts_eager_fallback.h>
+
+namespace torch {
+namespace lazy {
+
+std::atomic<bool>& in_lazy_mode() {
+    static std::atomic<bool> _in_lazy_mode{false};
+    return _in_lazy_mode;
+}
+
+void LazyModeEnter(c10::Device device) {
+    in_lazy_mode() = true;
+    // It is straightforward why we want to set the lazy key on entering the mode:
+    // we force operators (even on regular eager tensors) to route to lazy implementations
+    c10::impl::tls_set_dispatch_key_included(c10::DispatchKey::Lazy, true);
+}
+
+void LazyModeExit(c10::Device device) {
+    // Equally straightforward is that we no longer want the lazy key when we exit the mode:
+    // this lets operations on eager tensors outside the mode go back to normal eager behavior
+    c10::impl::tls_set_dispatch_key_included(c10::DispatchKey::Lazy, false);
+
+    // Less obvious is that we also set an 'unlazy' key on mode exit, which lets us specially handle
+    // any 'lazy' tensors that are alive after the mode exit.  This could be avoided if we can find another way
+    // to make lazy tensors interoperable with eager kernels.
+    // TODO(whc) PrivateUse1 is just for prototyping
+    c10::impl::tls_set_dispatch_key_included(c10::DispatchKey::PrivateUse1, true);
+
+    // At mode exit, we use the currently 'live' lazy tensors to define a graph to compile/execute
+    auto backend_device = torch::lazy::atenDeviceToBackendDevice(device);
+    auto backend_devices = {backend_device.toString()};
+    // wait=true: means we definitely submit all gpu work before exiting,
+    // does not sync on gpu completion
+    torch::lazy::LazyGraphExecutor::Get()->SyncLiveTensorsGraph(&backend_device, backend_devices, /*wait=*/true);
+
+    // Live lazy tensors should now all have eager tensors replacing their 'ir_value' fields, which can be
+    // accessed by eager kernels using the 'unlazy handler'
+    in_lazy_mode() = false;
```

Comment:
nitpick, would nested lazy mode be a thing?

If we don't want to support that, might as well disable it and assert on enter when we are already in a lazy context.

Reply:
> or so that it can be applied inside library code and then also around user code

This is a good argument for why we might want to allow nesting. Some library code may use Lazy Mode and then some user wants to call e.g. 3 libraries (one of which uses LazyMode) and then use Lazy Mode on the outside.

Though it's not too bad to say we don't support it for now, wait for users to ask for nesting, and then give it to them